kore/src/Kore/Builtin/Builtin.hs

```diff
@@ -239,22 +248,84 @@ type SymbolVerifiers = HashMap Text SymbolVerifier
   represention parameterized over @m@, the verification monad.
 
 -}
-type DomainValueVerifier child =
-       DomainValue Sort child -> Either (Error VerifyError) (Builtin child)
+newtype DomainValueVerifier patternType =
+    DomainValueVerifier
+        { runDomainValueVerifier
+            :: DomainValue Sort patternType
+            -> Either (Error VerifyError) (TermLikeF Variable patternType)
+        }
 
 -- | @DomainValueVerifiers@  associates a @DomainValueVerifier@ with each
 -- builtin
-type DomainValueVerifiers child = (HashMap Text (DomainValueVerifier child))
+type DomainValueVerifiers child = HashMap Text (DomainValueVerifier child)
+
+-- | Verify (and internalize) an application pattern.
+newtype ApplicationVerifier patternType =
+    ApplicationVerifier
+        { runApplicationVerifier
+            :: Application Symbol patternType
+            -> Either (Error VerifyError) (TermLikeF Variable patternType)
+        }
 
+{- | @SymbolKey@ names builtin functions and constructors.
+ -}
+data SymbolKey
+    = HookedSymbolKey !Text
+    -- ^ A builtin function identified by its @hook@ attribute.
+    | KlabelSymbolKey !Text
+    -- ^ A builtin constructor identified by its @klabel@ attribute.
+    deriving (Eq, Ord)
+    deriving (GHC.Generic)
+
+instance Hashable SymbolKey
+
+type ApplicationVerifiers patternType =
+    HashMap SymbolKey (ApplicationVerifier patternType)
+
+lookupApplicationVerifier
+    :: Symbol
+    -> ApplicationVerifiers patternType
+    -> Maybe (ApplicationVerifier patternType)
+lookupApplicationVerifier symbol verifiers = do
+    key <- getHook symbol <|> getKlabel symbol
+    HashMap.lookup key verifiers
+  where
+    getHook =
+        fmap HookedSymbolKey
+        . Attribute.Symbol.getHook . Attribute.Symbol.hook
+        . symbolAttributes
+    getKlabel =
+        fmap KlabelSymbolKey
+        . Attribute.Symbol.getKlabel . Attribute.Symbol.klabel
+        . symbolAttributes
 
 {- | Verify builtin sorts, symbols, and patterns.
  -}
 data Verifiers = Verifiers
     { sortDeclVerifiers    :: SortDeclVerifiers
     , symbolVerifiers      :: SymbolVerifiers
     , domainValueVerifiers :: DomainValueVerifiers Verified.Pattern
+    , applicationVerifiers :: ApplicationVerifiers Verified.Pattern
     }
 
+instance Semigroup Verifiers where
```

Comment:
It's not in `base` is it?

Reply:
Nope, it's in `semigroups` unfortunately, that's why I'm not sure it's worth adding a dependency for this. Maybe there are other `Semigroup`/`Monoid` instances and we could cut down more code? It might be worth it then. Perhaps we can just add a card to investigate?